core/types/calls.cc

```diff
@@ -899,13 +899,36 @@ DispatchResult dispatchCallSymbol(const GlobalState &gs, const DispatchArgs &arg
                 ait += numKwargs;
             }
         } else if (kwSplatIsHash) {
-            if (auto e = gs.beginError(args.callLoc(), errors::Infer::UntypedSplat)) {
-                e.setHeader("Passing a hash where the specific keys are unknown to a method taking keyword "
-                            "arguments");
-                auto &kwSplatArg = *aend;
-                auto kwSplatTPO = TypeAndOrigins{kwSplatType, kwSplatArg->origins};
-                e.addErrorSection(kwSplatTPO.explainGot(gs, args.originForUninitialized));
-                result.main.errors.emplace_back(e.build());
+            bool hasKwSplatParam = false;
+            bool hasKwParam = false;
+            const ArgInfo *kwSplatParam;
+            for (auto &param : data->arguments()) {
+                if (param.flags.isKeyword && param.flags.isRepeated) {
+                    kwSplatParam = &param;
+                    hasKwSplatParam = true;
+                } else if (param.flags.isKeyword && !param.flags.isRepeated) {
+                    hasKwParam = true;
+                }
+            }
+
+            auto &kwSplatArg = *aend;
+            auto kwSplatTPO = TypeAndOrigins{kwSplatType, kwSplatArg->origins};
+
+            if (hasKwSplatParam && !hasKwParam) {
+                if (auto e = matchArgType(gs, *constr, args.callLoc(), args.receiverLoc(), symbol, method, kwSplatTPO,
+                                          *kwSplatParam, args.selfType, targs, kwargsLoc, args.originForUninitialized,
+                                          args.args.size() == 1)) {
+                    result.main.errors.emplace_back(std::move(e));
+                }
+            } else {
+                // A keyword splat was passed in, but none of the declared arguments are keyword splats,
+                // or there is a keyword splat argument, but there is one ore more keyword non-splat argument
+                if (auto e = gs.beginError(args.callLoc(), errors::Infer::UntypedSplat)) {
+                    e.setHeader("Passing a hash where the specific keys are unknown to a method taking keyword "
+                                "arguments");
```

Comment:
Do you think it's worth making this error message more specific? e.g. this comment has an "either/or" which is very helpful, but neither sides of the either/or end up in the message that we show to the user.

Reply:
I can't think of any obvious ways to make it more helpful. In both cases, the method will have declared keyword arguments. In one of the cases, there's also a declared keyword splat argument, but the reason for the error in both cases are the keyword arguments, not the keyword splat.